Add per_process_memory_fraction to PYTORCH_CUDA_ALLOC_CONF

lakshayg opened the pull request:
torch.cuda.memory.set_per_process_memory_fraction allows setting
an upper bound on how much device memory is allocated. This PR
exposes this setting to an environment variable.

For example, PYTORCH_CUDA_ALLOC_CONF="per_process_memory_fraction:0.5"
will limit the device memory to half of the available memory.

cc @ptrblck @msaroufim @eqy @jerryzh168

eqy commented on test/test_cuda.py:
```diff
@@ -4517,6 +4517,21 @@ def power2_div(size, div_factor):
 
+        device = torch._C._cuda_getDevice()
+        torch.cuda.memory._set_allocator_settings("per_process_memory_fraction:0.5")
+        self.assertEqual(torch.cuda.memory.get_per_process_memory_fraction(device), 0.5)
```
Hmm, what happens when the `torch.cuda.memory.set_per_process_memory_fraction` API is used?

eqy commented on c10/cuda/CUDACachingAllocator.h:
```diff
@@ -208,7 +209,15 @@ class CUDAAllocator : public DeviceAllocator {
-  virtual void init(int device_count) = 0;
+
+  // Any classes overriding init should call CUDAAllocator::init
+  virtual void init(int device_count) {
```
Will this always run even if it is a no-op? e.g., setting memory fraction to 1.0?

Also curious if somehow it is changing the current CUDA device to the last one

ngimel commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -1183,6 +1180,8 @@ class DeviceCachingAllocator {
   // device statistics
   DeviceStats stats;
 
+  c10::DeviceIndex device_id;
```
separate changes related to adding `device_id` as a new member from changes related to per_process_memory_fraction, for ease of reviewing

lakshayg committed:
Add per_process_memory_fraction to PYTORCH_CUDA_ALLOC_CONF

torch.cuda.memory.set_per_process_memory_fraction allows setting
an upper bound on how much device memory is allocated. This PR
exposes this setting to an environment variable.

For example, PYTORCH_CUDA_ALLOC_CONF="per_process_memory_fraction:0.5"
will limit the device memory to half of the available memory.

lakshayg committed:
Remove support from CUDAMallocAsyncAllocator

Async allocator does not support other config options either

lakshayg commented:
@ngimel I changed the PR and removed the support for the CUDAMallocAsyncAllocator.

This change was motivated by a test failure in the CUDAMallocAsyncAllocator. The test was failing because that allocator was trying to set the memory fraction at init time. Based on the comment on CUDAMallocAsyncAllocator::init, that function should not call anything that inits the CUDA context and setting the memory fraction requires doing that.

Instead of trying to support it properly, I noticed that the async allocator does not support other config options either so it's probably not worth supporting this option either. Please let me know your thought and re-approve it you agree!

eqy reviewed (approved):
should there be a warning if this setting is attempted to be used with the cudaMallocAsync allocator?

lakshayg commented:
@pytorchmergebot merge
